File: android-kit-base/src/test/java/com/mparticle/mock/MockKitConfiguration.java
package com.mparticle.mock;

import android.util.SparseBooleanArray;

import com.mparticle.internal.Logger;
import com.mparticle.kits.KitConfiguration;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

/**
 * This class is necessary b/c SparseBooleanArray is not available while unit testing.
 */
public class MockKitConfiguration extends KitConfiguration {

    @Override
    public KitConfiguration parseConfiguration(JSONObject json) throws JSONException {
        mTypeFilters = new MockSparseBooleanArray();
        mNameFilters = new MockSparseBooleanArray();
        mAttributeFilters = new MockSparseBooleanArray();
        mScreenNameFilters = new MockSparseBooleanArray();
        mScreenAttributeFilters = new MockSparseBooleanArray();
        mUserIdentityFilters = new MockSparseBooleanArray();
        mUserAttributeFilters = new MockSparseBooleanArray();
        mCommerceAttributeFilters = new MockSparseBooleanArray();
        mCommerceEntityFilters = new MockSparseBooleanArray();
        return super.parseConfiguration(json);
    }

    public static KitConfiguration createKitConfiguration(JSONObject json) throws JSONException{
        return new MockKitConfiguration().parseConfiguration(json);
    }

    public static KitConfiguration createKitConfiguration() throws JSONException{
        JSONObject jsonObject = new JSONObject();
        jsonObject.put("id", 42);
        return new MockKitConfiguration().parseConfiguration(jsonObject);
    }

    @Override
    protected SparseBooleanArray convertToSparseArray(JSONObject json) {
        SparseBooleanArray map = new MockSparseBooleanArray();
        for (Iterator<String> iterator = json.keys(); iterator.hasNext();) {
            try {
                String key = iterator.next();
                map.put(Integer.parseInt(key), json.getInt(key) == 1);
            }catch (JSONException jse){
                Logger.error("Issue while parsing kit configuration: " + jse.getMessage());
            }
        }
        return map;
    }
    class MockSparseBooleanArray extends SparseBooleanArray {
        @Override
        public boolean get(int key) {
            return get(key, false);
        }

        @Override
        public boolean get(int key, boolean valueIfKeyNotFound) {
            System.out.print("SparseArray getting: " + key);
            if (map.containsKey(key)) {
                return map.get(key);
            }else{
                return valueIfKeyNotFound;
            }
        }

        Map<Integer, Boolean> map = new HashMap<Integer, Boolean>();
        @Override
        public void put(int key, boolean value) {
            map.put(key, value);
        }

        @Override
        public void clear() {
            map.clear();
        }

        @Override
        public int size() {
            return map.size();
        }

        @Override
        public String toString() {
            return map.toString();
        }
    }
}

File: kits/rokt-kit/src/test/kotlin/com/mparticle/kits/mocks/MockKitConfiguration.kt
package com.mparticle.kits.mocks

import android.util.SparseBooleanArray
import com.mparticle.internal.Logger
import com.mparticle.kits.KitConfiguration
import org.json.JSONException
import org.json.JSONObject

open class MockKitConfiguration : KitConfiguration() {
    @Throws(JSONException::class)
    override fun parseConfiguration(json: JSONObject): KitConfiguration {
        mTypeFilters = MockSparseBooleanArray()
        mNameFilters = MockSparseBooleanArray()
        mAttributeFilters = MockSparseBooleanArray()
        mScreenNameFilters = MockSparseBooleanArray()
        mScreenAttributeFilters = MockSparseBooleanArray()
        mUserIdentityFilters = MockSparseBooleanArray()
        mUserAttributeFilters = MockSparseBooleanArray()
        mCommerceAttributeFilters = MockSparseBooleanArray()
        mCommerceEntityFilters = MockSparseBooleanArray()
        return super.parseConfiguration(json)
    }

    override fun convertToSparseArray(json: JSONObject): SparseBooleanArray {
        val map: SparseBooleanArray = MockSparseBooleanArray()
        val iterator = json.keys()
        while (iterator.hasNext()) {
            try {
                val key = iterator.next().toString()
                map.put(key.toInt(), json.getInt(key) == 1)
            } catch (jse: JSONException) {
                Logger.error("Issue while parsing kit configuration: " + jse.message)
            }
        }
        return map
    }

    internal inner class MockSparseBooleanArray : SparseBooleanArray() {
        override fun get(key: Int): Boolean = get(key, false)

        override fun get(key: Int, valueIfKeyNotFound: Boolean): Boolean {
            print("SparseArray getting: $key")
            return if (map.containsKey(key)) {
                map[key] ?: valueIfKeyNotFound
            } else {
                valueIfKeyNotFound
            }
        }

        var map: MutableMap<Int, Boolean> = HashMap()
        override fun put(key: Int, value: Boolean) {
            map[key] = value
        }

        override fun clear() {
            map.clear()
        }

        override fun size(): Int = map.size

        override fun toString(): String = map.toString()
    }

    companion object {
        @Throws(JSONException::class)
        fun createKitConfiguration(json: JSONObject): KitConfiguration = MockKitConfiguration().parseConfiguration(json)

        @Throws(JSONException::class)
        fun createKitConfiguration(): KitConfiguration {
            val jsonObject = JSONObject()
            jsonObject.put("id", 42)
            return MockKitConfiguration().parseConfiguration(jsonObject)
        }
    }
}
